File: c++/242.valid-anagram.cpp
/*
 * @lc app=leetcode id=242 lang=cpp
 *
 * [242] Valid Anagram
 */
#include <string>
using std::string;

// @lc code=start
class Solution {
public:

	// something like this, will probably come back for this solution:
	// rn failes test 35. 
    // bool isAnagram(string s, string t) {
	// 	// lets do a random hash(?)
	//   	if (s.size() != t.size()) return false;
	// 	int sHash = 0;
	// 	int tHash = 0;
	// 	for (int i = 0; i < s.size(); i++) {
	// 		sHash = sHash + s[i] * s[i] ;
	// 		tHash = tHash + t[i] * t[i] ;
	// 	}
	// 	return sHash == tHash;
    // }

// Accepted
// 36/36 cases passed (3 ms)
// Your runtime beats 98.44 % of cpp submissions
// Your memory usage beats 8.71 % of cpp submissions (7.4 MB)
	#define ALPHABET_LENGTH 26
    bool isAnagram(string s, string t) {
	 	if (s.size() != t.size()) return false;
		int count[ALPHABET_LENGTH] = {0};
		for (int i = 0; i < s.size(); i++){
			count[s[i] - (int)'a']++;
			count[t[i] - (int)'a']--;
		}
		for (auto i : count) {
			if (i) return false;
		}
		return true;
    }
// Accepted
// 36/36 cases passed (863 ms)
// Your runtime beats 5.01 % of cpp submissions
// Your memory usage beats 46.76 % of cpp submissions (7.3 MB)
    // bool isAnagram(string s, string t) {
	// 	if (s.size() != t.size()) return false;
	// 	// dumb solution:
	// 	for (auto i:s){
	// 		size_t index = t.find_first_of(i);
	// 		if (index == -1) return false;
	// 		t.erase(index, 1);
	// 	}
	// 	return t.empty();
    // }
};
// @lc code=end

File: python/242.valid-anagram.py
#
# @lc app=leetcode id=242 lang=python3
#
# [242] Valid Anagram
#

# My naive python solution. Will be looking at the discussion.
# Accepted
# 36/36 cases passed (110 ms)
# Your runtime beats 10.5 % of python3 submissions
# Your memory usage beats 66.36 % of python3 submissions (14.4 MB)
# @lc code=start
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(s) :
            return False
        
        ascii_map = [0] * 26 
        for c in s:
            ascii_map[ord(c) - ord('a')]+=1

        for c in t:
            ascii_map[ord(c) - ord('a')]-=1

        return not any(ascii_map)
        # for c in ascii_map:
        #     if c != 0:
        #         return False
        # return True
# @lc code=end

